### PR TITLE
[JENKINS-41125] Update events to a branch with a name that gets mangled do not get processed

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-41125](https://issues.jenkins-ci.org/browse/JENKINS-41125)

@reviewbybees

### CODE REVIEW

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 1]
```diff
@@ -1409,7 +1417,7 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
                     for (SCMSource source : p.getSCMSources()) {
                         if (event.isMatch(source)) {
                             for (SCMHead h : event.heads(source).keySet()) {
-                                if (p.getItem(h.getName()) == null) {
+                                if (p.getItemByBranchName(h.getName()) == null) {
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Not strictly necessary as getItem will pick up the old name, but removes the 1 in 33,554,432 chance if somebody for example has a branch called `feature/ux-1` and a branch called `feature-ux-1.0gorp7`

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 2]
```diff
@@ -1323,13 +1330,14 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
                                             // we have already seen
                                             LOGGER.log(Level.FINE,
                                                     "{0} {1} {2,date} {2,time}: Checking {3} -> Ignoring event as "
-                                                            + "revision {4} is same as last build of {5}",
+                                                            + "revision {4} is same as last build of {5} (job {6})",
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 3]
```diff
@@ -1036,7 +1036,7 @@ private int processHeadCreate(SCMHeadEvent<?> event, TaskListener global, String
                         );
                         for (SCMHead h : event.heads(source).keySet()) {
                             String name = h.getName();
-                            Job job = p.getItem(name);
+                            Job job = p.getItemByBranchName(name);
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Not strictly necessary as getItem will pick up the old name, but removes the 1 in 33,554,432 chance if somebody for example has a branch called `feature/ux-1` and a branch called `feature-ux-1.0gorp7`

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 4]
```diff
@@ -1168,8 +1168,8 @@ private int processHeadCreate(SCMHeadEvent<?> event, TaskListener global, String
             return matchCount;
         }
 
-        private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String eventClass, String eventType, long eventTimestamp,
-                                      int matchCount) {
+        private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String eventClass, String eventType,
+                                      long eventTimestamp, int matchCount) {
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Reformatting

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 5]
```diff
@@ -1196,15 +1196,17 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                 jobs.clear();
                 for (Job i : p.getItems()) {
-                    if (!candidateNames.contains(i.getName())) {
-                        continue;
-                    }
                     if (_factory.isProject(i)) {
                         Branch branch = _factory.getBranch(i);
+                        if (!candidateNames.contains(branch.getName())) {
+                            continue;
+                        }
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: This is part of the actual fix. We need to move the check against candidate names to run against the branch name not the job name

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 6]
```diff
@@ -1196,15 +1196,17 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                         if (branch instanceof Branch.Dead) {
-                            LOGGER.log(Level.FINEST, "{0} {1} {2,date} {2,time}: Checking {3} -> Resurrect dead branch {4} ?",
+                            LOGGER.log(Level.FINEST, "{0} {1} {2,date} {2,time}: Checking {3} -> Resurrect dead "
+                                            + "branch {4} (job {5})?",
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 7]
```diff
@@ -1196,15 +1196,17 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                     new Object[]{
-                                            eventClass, eventType, eventTimestamp, pFullName, i.getName()
+                                            eventClass, eventType, eventTimestamp, pFullName, branch.getName(),
+                                            i.getName()
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 8]
```diff
@@ -1228,9 +1230,11 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
                         } else {
                             // TODO takeover
                             LOGGER.log(Level.FINEST,
-                                    "{0} {1} {2,date} {2,time}: Checking {3} -> Matches existing branch {4} ?",
+                                    "{0} {1} {2,date} {2,time}: Checking {3} -> Matches existing branch {4} "
+                                            + "(job {5})?",
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 9]
```diff
@@ -1228,9 +1230,11 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                     new Object[]{
-                                            eventClass, eventType, eventTimestamp, pFullName, i.getName()
+                                            eventClass, eventType, eventTimestamp, pFullName, branch.getName(),
+                                            i.getName()
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 10]
```diff
@@ -1244,9 +1248,10 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
                                 // to see if this is a takeover
                                 LOGGER.log(Level.FINEST,
                                         "{0} {1} {2,date} {2,time}: Checking {3} -> Event does not match current "
-                                                + "source of {4}, checking for take-over",
+                                                + "source of {4} (job {5}), checking for take-over",
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 11]
```diff
@@ -1244,9 +1248,10 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                         new Object[]{
-                                                eventClass, eventType, eventTimestamp, pFullName, i.getName()
+                                                eventClass, eventType, eventTimestamp, pFullName, branch.getName(),
+                                                i.getName()
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 12]
```diff
@@ -1261,7 +1266,7 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
                                     if (ourPriority > priority && oldPriority > priority && rMap != null) {
                                         // only need to check for takeover when the event is higher priority
                                         for (SCMHead h: rMap.keySet()) {
-                                            if (i.getName().equals(h.getName())) {
+                                            if (branch.getName().equals(h.getName())) {
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: This is part of the fix, need to compare the branch name

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 13]
```diff
@@ -1276,20 +1281,20 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
                                 if (oldPriority < ourPriority) {
                                     LOGGER.log(Level.FINEST,
                                             "{0} {1} {2,date} {2,time}: Checking {3} -> Ignoring event for {4} "
-                                                    + "from source #{5} as source #{6} owns the branch name",
+                                                    + "(job {5}) from source #{6} as source #{7} owns the branch name",
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 14]
```diff
@@ -1276,20 +1281,20 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                             new Object[]{
-                                                    eventClass, eventType, eventTimestamp, pFullName, i.getName(),
-                                                    ourPriority, oldPriority
+                                                    eventClass, eventType, eventTimestamp, pFullName, branch.getName(),
+                                                    i.getName(), ourPriority, oldPriority
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 15]
```diff
@@ -1276,20 +1281,20 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                             }
                                     );
                                     continue;
                                 } else {
                                     LOGGER.log(Level.FINER,
                                             "{0} {1} {2,date} {2,time}: Checking {3} -> Takeover event for {4} "
-                                                    + "by source #{5} from source #{6}",
+                                                    + "(job {5}) by source #{5} from source #{6}",
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 16]
```diff
@@ -1276,20 +1281,20 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                             new Object[]{
-                                                    eventClass, eventType, eventTimestamp, pFullName, i.getName(),
-                                                    ourPriority, oldPriority
+                                                    eventClass, eventType, eventTimestamp, pFullName, branch.getName(),
+                                                    i.getName(), ourPriority, oldPriority
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 17]
```diff
@@ -1309,9 +1314,11 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
                             }
                             if (match) {
                                 LOGGER.log(Level.FINE,
-                                        "{0} {1} {2,date} {2,time}: Checking {3} -> Event matches source of {4}",
+                                        "{0} {1} {2,date} {2,time}: Checking {3} -> Event matches source of {4} "
+                                                + "(job {5})",
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 18]
```diff
@@ -1309,9 +1314,11 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                         new Object[]{
-                                                eventClass, eventType, eventTimestamp, pFullName, i.getName()
+                                                eventClass, eventType, eventTimestamp, pFullName, branch.getName(),
+                                                i.getName()
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

File: src/main/java/jenkins/branch/MultiBranchProject.java [hunk 19]
```diff
@@ -1323,13 +1330,14 @@ private int processHeadUpdate(SCMHeadEvent<?> event, TaskListener global, String
...
                                                     "{0} {1} {2,date} {2,time}: Checking {3} -> Ignoring event as "
-                                                            + "revision {4} is same as last build of {5}",
+                                                            + "revision {4} is same as last build of {5} (job {6})",
                                                     new Object[]{
                                                             eventClass,
                                                             eventType,
                                                             eventTimestamp,
                                                             pFullName,
                                                             revision,
+                                                            branch.getName(),
```
Comment (stephenc) [inline src/main/java/jenkins/branch/MultiBranchProject.java]: Logging

**Comment by aheritier** [comment]
ok AFAICT 🐝

File: src/test/java/integration/EventsTest.java [hunk 20]
```diff
@@ -79,6 +79,24 @@
...
     /**
+     * A branch name that contains a slash
+     */
+    private static final String SLASHY_BRANCH_NAME = "feature/ux-1";
+    /**
+     * The mangled name of {@link #SLASHY_BRANCH_NAME}
+     */
+    private static final String SLASHY_JOB_NAME = "feature-ux-1.0gorp7";
+    /**
+     * A branch name with unicode characters that have a two canonical forms (Korean for New Features).
```
Comment (jglick) [inline src/test/java/integration/EventsTest.java]: Why do we care about canonical forms? My understanding was that we only paid attention to the UTF-8 representation of a branch name as supplied by the SCM source.

File: src/test/java/integration/EventsTest.java [hunk 21]
```diff
@@ -464,6 +482,50 @@ public void given_multibranchWithSources_when_matchingEvent_then_branchesAreFoun
     }
 
     @Test
+    public void given_multibranchWithSlashySources_when_matchingEvent_then_branchesAreFoundAndBuilt() throws Exception {
```
Comment (jglick) [inline src/test/java/integration/EventsTest.java]:
```java
@Issue("JENKINS-41125")
```

File: src/test/java/integration/EventsTest.java [hunk 22]
```diff
@@ -464,6 +482,50 @@ public void given_multibranchWithSources_when_matchingEvent_then_branchesAreFoun
...
+        try (MockSCMController c = MockSCMController.create()) {
+            c.createRepository("foo");
+            c.createBranch("foo", SLASHY_BRANCH_NAME);
+            c.deleteBranch("foo", "master");
+            BasicMultiBranchProject prj = r.jenkins.createProject(BasicMultiBranchProject.class, "foo");
+            prj.setCriteria(null);
```
Comment (jglick) [inline src/test/java/integration/EventsTest.java]: What does this do?

File: src/test/java/integration/EventsTest.java (hunk 20 above)
Comment (stephenc) [inline src/test/java/integration/EventsTest.java]: we don't care about canonical forms but this has the advantage of checking that we don't care

File: src/test/java/integration/EventsTest.java (hunk 22 above)
Comment (stephenc) [inline src/test/java/integration/EventsTest.java]: it means that all branches are valid. no need to check them for marker files

**Pull request merged**